### PR TITLE
Support new resource name annotations

### PR DESCRIPTION
**Pull request by michaelbausor**
WIP, but relevant changes:
- Determine when a resource type is references via a child_type reference. This indicates that we need to generate a resource name type for the parent of that type.
- Unique entity name resolution via a trie structure based on the variable segments of the pattern (same as what is done in the PR to gapic-generator)
- Completely ignore field_name_patterns - I don't think they are used at all by the rest of the plugin?

An example to motivate the complex name resolution: BQDT TransferRun: https://github.com/michaelbausor/googleapis/blob/resources-redux/google/cloud/bigquery/datatransfer/v1/transfer.proto#L117

### CODE REVIEW

**Comment by michaelbausor** [comment]
@lukesneeringer @andreamlin PTAL. Note changes in baselines - these were made so that the v1 config and v2 config tests could share baselines. Places that it wasn't possible to get full compatibility, requiring changes, were the naming of variable segments (e.g. archive vs archived, affecting name of generated resource name class)

File: test/testdata/library_gapic.yaml [hunk 1]
```diff
@@ -1,29 +1,30 @@
...
-- name_pattern: shelves/{shelf_id}/books/{return} # Test using a Java keyword ("return").
+- name_pattern: shelves/{shelf}/books/{book}
   entity_name: book
   language_overrides:
     - language: java
       entity_name: shelf_book
-- name_pattern: archives/{archive_id}/books/{book_id=**}
-  entity_name: archived_book
+- name_pattern: archives/{archive}/books/{book}
+  entity_name: archive_book
```
Comment (andreamlin) [inline test/testdata/library_gapic.yaml]: Could we have a language_override for java for `archive_book` to `archived_book` so there is no diff, or a smaller one, in the baseline file?

File: test/testdata/java_shelf_name.baseline [hunk 2]
```diff
@@ -114,28 +114,28 @@ public class ShelfName implements ResourceName {
...
 
   /** Builder for ShelfName. */
   public static class Builder {
 
-    private String shelfId;
+    private String shelf;
 
-    public String getShelfId() {
-      return shelfId;
+    public String getShelf() {
```
Comment (andreamlin) [inline test/testdata/java_shelf_name.baseline]: Are we forced to change the `{shelf_id}` path substring into `{shelf}`? It changes this public method name.

Comment (michaelbausor) [reply test/testdata/java_shelf_name.baseline]: Yes - if we don't, it changes the inferred class name (e.g. you get ShelfIdName). This change is actually what we want to be consistent with most (all??) APIs in googleapis.

File: test/testdata/library_gapic.yaml (hunk 1 above)
Comment (michaelbausor) [reply test/testdata/library_gapic.yaml]: Just looked into this, but the problem is then there is a difference between the v1 and v2 config cases, and the goal was to try to make them use the same set of baseline filed.

File: test/testdata/java_shelf_name.baseline (hunk 2 above)
Comment (andreamlin) [reply test/testdata/java_shelf_name.baseline]: Ok, this is a breaking change for the GA Firestore client. While i think the impact is very small, I'm not sure how we should handle versioning that client.

File: test/testdata/library_gapic.yaml (hunk 1 above)
Comment (michaelbausor) [reply test/testdata/library_gapic.yaml]: Actually, I did make this work by adding it to the v2 config as well.

Comment (andreamlin) [reply test/testdata/library_gapic.yaml]: ok, so long as it is possible

File: test/testdata/java_shelf_name.baseline (hunk 2 above)
Comment (michaelbausor) [reply test/testdata/java_shelf_name.baseline]: Maybe we can work around it with entity_name overrides?

**Pull request merged**